### PR TITLE
MONGOID-5386 Pass through uncastable values in evolve

### PR DESCRIPTION
**Pull request by Neilshweky**

### CODE REVIEW

File: lib/mongoid/criteria/queryable/extensions/boolean.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ module ClassMethods
             # @return [ true, false ] The boolean value.
             def evolve(object)
               __evolve__(object) do |obj|
-                obj.to_s =~ (/\A(true|t|yes|y|on|1|1.0)\z/i) ? true : false
+                mongoize(obj)
```
Comment (johnnyshields) [inline lib/mongoid/criteria/queryable/extensions/boolean.rb:23]: This doesn't pass thru uncastable?

How about:

```ruby
return nil if obj.nil?

mongoize(obj) || obj
```

Comment (johnnyshields) [reply lib/mongoid/criteria/queryable/extensions/boolean.rb:23]: Roger that, thanks!

File: lib/mongoid/criteria/queryable/extensions/symbol.rb [hunk 2]
```diff
@@ -59,7 +59,9 @@ def add_key(name, strategy, operator, additional = nil, &block)
             #
             # @return [ Symbol ] The value as a symbol.
             def evolve(object)
-              __evolve__(object) { |obj| obj.to_sym }
+              __evolve__(object) do |obj|
+                obj.try(:to_sym)
```
Comment (p-mongo) [inline lib/mongoid/criteria/queryable/extensions/symbol.rb:63]: The `try` call is the same as `obj&.to_sym` right?

**Review by p-mongo — APPROVED** [review]

File: lib/mongoid/criteria/queryable/extensions/symbol.rb (hunk 2 above)
Comment (Neilshweky) [reply lib/mongoid/criteria/queryable/extensions/symbol.rb:63]: I don't think so... iirc the `&.` safe-navigation operator either calls the method if the value is not nil, or returns nil if it is.

> What I'm trying to protect against here is the case where the to_sym method doesn't exist on the value (for example, arrays). Changing to the safe navigation operator should cause that case to raise a NoMethodError

Comment (p-mongo) [reply lib/mongoid/criteria/queryable/extensions/symbol.rb:63]: Aren't they the same?

```irb(main):001:0> [].try(:to_sym)
(irb):1:in `<main>': undefined method `try' for []:Array (NoMethodError)
	from /usr/lib/ruby/gems/3.0.0/gems/irb-1.3.5/exe/irb:11:in `<top (required)>'
	from /usr/bin/irb:23:in `load'
	from /usr/bin/irb:23:in `<main>'

```

Comment (Neilshweky) [reply lib/mongoid/criteria/queryable/extensions/symbol.rb:63]: Try this again with rails loaded:
```
irb(main):001:0> [].try(:to_sym)
=> nil
irb(main):002:0> []&.to_sym
Traceback (most recent call last):
       16: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/friendly_errors.rb:103:in `with_friendly_errors'
       15: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/exe/bundle:48:in `block in <top (required)>'
       14: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli.rb:25:in `start'
       13: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
       12: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli.rb:31:in `dispatch'
       11: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
       10: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
        9: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
        8: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli.rb:484:in `exec'
        7: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli/exec.rb:23:in `run'
        6: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli/exec.rb:58:in `kernel_load'
        5: from /Users/neil.shweky/.gem/gems/bundler-2.3.6/lib/bundler/cli/exec.rb:58:in `load'
        4: from /Users/neil.shweky/.gem/bin/irb:23:in `<top (required)>'
        3: from /Users/neil.shweky/.gem/bin/irb:23:in `load'
        2: from /Users/neil.shweky/.rbenv/versions/2.7.5/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        1: from (irb):2
NoMethodError (undefined method `to_sym' for []:Array)
Did you mean?  to_s
               to_set
```

Comment (p-mongo) [reply lib/mongoid/criteria/queryable/extensions/symbol.rb:63]: I see.

**Pull request merged**